### PR TITLE
add feature 'product white list'

### PR DESCRIPTION
**Pull request by linr211**
add feature 'product white list' to validate a product from request in the interceptor, if the product is not in the white list then throw the error message to the response body.

### CODE REVIEW

File: g11n-ws/modules/md-service-authen/src/main/java/com/vmware/vip/core/validation/ParameterValidation.java [hunk 1]
```diff
@@ -17,6 +21,13 @@
 
 public class ParameterValidation implements IVlidation {
 	HttpServletRequest request = null;
+	public final static String TAG_BUNDLE_BASE_PATH = "bundleBasedPath";
+	public final static String BUNDLE_FILE = "bundle.json";
+
+	/**
+	 * A cached map to store the data of product white list
+	 */
+	private  static final Map<String, Object> MAP_PRODUCTS_VERSIONS = new HashMap<>();
```
Comment (huihuiw01) [inline g11n-ws/modules/md-service-authen/src/main/java/com/vmware/vip/core/validation/ParameterValidation.java]: Can this white list be updated at runtime?

File: g11n-ws/modules/md-service-authen/src/main/java/com/vmware/vip/core/validation/ParameterValidation.java [hunk 2]
```diff
@@ -17,6 +21,13 @@
 
 public class ParameterValidation implements IVlidation {
 	HttpServletRequest request = null;
+	public final static String TAG_BUNDLE_BASE_PATH = "bundleBasedPath";
+	public final static String BUNDLE_FILE = "bundle.json";
```
Comment (huihuiw01) [inline g11n-ws/modules/md-service-authen/src/main/java/com/vmware/vip/core/validation/ParameterValidation.java]: Looks like this PR doesn't include code for generating this file.

File: g11n-ws/vip-manager-i18n/src/main/java/com/vmware/vip/core/conf/WebConfiguration.java [hunk 3]
```diff
@@ -77,8 +77,10 @@
 
 	@Value("${vipservice.cross.domain.maxage}")
 	private String maxAge;
-	
-	
+
+	@Value("${translation.bundle.file.basepath}")
```
Comment (huihuiw01) [inline g11n-ws/vip-manager-i18n/src/main/java/com/vmware/vip/core/conf/WebConfiguration.java]: Should this parameter be added in application.properties? I don't see any change for application.properties.

File: g11n-ws/modules/md-service-authen/src/main/java/com/vmware/vip/core/validation/ParameterValidation.java (hunk 1 above)
Comment (linr211) [reply g11n-ws/modules/md-service-authen/src/main/java/com/vmware/vip/core/validation/ParameterValidation.java]: it requires to reboot vip service to reload the cache.

File: g11n-ws/modules/md-service-authen/src/main/java/com/vmware/vip/core/validation/ParameterValidation.java (hunk 2 above)
Comment (linr211) [reply g11n-ws/modules/md-service-authen/src/main/java/com/vmware/vip/core/validation/ParameterValidation.java]: it's done by build.gradle in translation.git project.

File: g11n-ws/vip-manager-i18n/src/main/java/com/vmware/vip/core/conf/WebConfiguration.java (hunk 3 above)
Comment (linr211) [reply g11n-ws/vip-manager-i18n/src/main/java/com/vmware/vip/core/conf/WebConfiguration.java]: This parameter exists in application.properties already, here is just to read it.

File: g11n-ws/modules/md-service-authen/src/main/java/com/vmware/vip/core/validation/ParameterValidation.java [hunk 4]
```diff
@@ -202,4 +216,23 @@ private void validatePattern(HttpServletRequest request)
...
+     * validate the product name by the white list
+     *
+     * @param productName
+     * @param whiteListFilePath
+     * @throws ValidationException
+     */
+    private void validateProductByWhiteList(String productName, String whiteListFilePath) throws ValidationException {
+        if (ParameterValidation.MAP_PRODUCTS_VERSIONS.isEmpty()) {
+            Map<String, Object> m = JSONUtils.getMapFromJsonFile(whiteListFilePath);
+            ParameterValidation.MAP_PRODUCTS_VERSIONS.putAll(m);
```
Comment (Xiaochao8) [inline g11n-ws/modules/md-service-authen/src/main/java/com/vmware/vip/core/validation/ParameterValidation.java]: How about assing m to ParameterValidation.MAP_PRODUCTS_VERSIONS directly?

File: g11n-ws/modules/md-service-authen/src/main/java/com/vmware/vip/core/validation/ParameterValidation.java [hunk 5]
```diff
@@ -202,4 +216,23 @@ private void validatePattern(HttpServletRequest request)
...
+        }
+        if (!ParameterValidation.MAP_PRODUCTS_VERSIONS.isEmpty()) {
+            if (!ParameterValidation.MAP_PRODUCTS_VERSIONS.containsKey(productName)) {
```
Comment (Xiaochao8) [inline g11n-ws/modules/md-service-authen/src/main/java/com/vmware/vip/core/validation/ParameterValidation.java]: Merge 2 if into 1.

**Pull request merged**